tardis/rest/app/security.py

```diff
@@ -1,127 +1,119 @@
+class DatabaseUser(BaseUser):
+    hashed_password: str
 
-    try:
-        payload = jwt.decode(token, get_secret_key(), algorithms=[get_algorithm()])
-        user_name: str = payload.get("sub")
-        token_scopes = payload.get("scopes", [])
-        token_data = TokenData(scopes=token_scopes, user_name=user_name)
-    except (JWTError, ValidationError) as err:
-        raise HTTPException(
-            status_code=status.HTTP_401_UNAUTHORIZED,
-            detail="Could not validate credentials",
-            headers={"WWW-Authenticate": authenticate_value},
-        ) from err
 
-    for scope in security_scopes.scopes:
-        if scope not in token_data.scopes:
+def check_scope_permissions(requested_scopes: List[str], allowed_scopes: List[str]):
+    for scope in requested_scopes:
+        if scope not in allowed_scopes:
             raise HTTPException(
-                status_code=status.HTTP_401_UNAUTHORIZED,
-                detail="Not enough permissions",
-                headers={"WWW-Authenticate": authenticate_value},
+                status_code=status.HTTP_403_FORBIDDEN,
+                detail={
+                    "msg": "Not enough permissions",
+                    "scopesRequired": requested_scopes,
+                },
             ) from None
 
-    return token_data
+
+def check_authorization(
+    security_scopes: SecurityScopes, Authorize: AuthJWT = Depends()
+) -> AuthJWT:
+    # No authorization without authentication
+    Authorize.jwt_required()
+
+    token_scopes = get_token_scopes(Authorize)
+    check_scope_permissions(security_scopes.scopes, token_scopes)
+
+    return Authorize
 
 
-def check_authentication(user_name: str, password: str) -> UserCredentials:
+def check_authentication(user_name: str, password: str) -> DatabaseUser:
     user = get_user(user_name)
+
     if not user:
-        raise HTTPException(status_code=400, detail="Incorrect username or password")
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Incorrect username or password",
+        )
+
     if checkpw(password.encode(), user.hashed_password.encode()):
         return user
     else:
-        raise HTTPException(status_code=400, detail="Incorrect username or password")
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Incorrect username or password",
+        )
 
 
-@lru_cache(maxsize=1)
-def get_algorithm() -> str:
+def get_token_scopes(Authorize: AuthJWT) -> List[str]:
     try:
-        rest_service = Configuration().Services.restapi
-    except AttributeError:
-        raise TardisError(
-            "TARDIS RestService not configured while accessing algorithm!"
-        ) from None
-    else:
-        return rest_service.algorithm
+        token_scopes: List[str] = Authorize.get_raw_jwt()["scopes"]
+    except KeyError:
+        raise TardisError("Scopes not defined in jwt token") from None
```

Comment:
Raising `TardisError` will crash the entire application, not only the REST interface? Is this intended behaviour here? Can somebody trigger this by providing a malicious token?